contracts/token/ERC721_Enumerable_base.cairo

```diff
@@ -0,0 +1,247 @@
+    ERC721_burn(token_id)
+    return ()
+end
+
+func ERC721_Enumerable_transferFrom{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(_from: felt, to: felt, token_id: Uint256):
+    _remove_token_from_owner_enumeration(_from, token_id)
+    _add_token_to_owner_enumeration(to, token_id)
+    ERC721_transferFrom(_from, to, token_id)
+    return ()
+end
+
+func ERC721_Enumerable_safeTransferFrom{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(
+        _from: felt, 
+        to: felt, 
+        token_id: Uint256, 
+        data_len: felt,
+        data: felt*
+    ):
+    _remove_token_from_owner_enumeration(_from, token_id)
+    _add_token_to_owner_enumeration(to, token_id)
+    ERC721_safeTransferFrom(_from, to, token_id, data_len, data)
+    return ()
+end
+
+#
+# Internals
+#
+
+func _add_token_to_all_tokens_enumeration{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(token_id: Uint256):
+    alloc_locals
+    let (supply: Uint256) = all_tokens_len.read()
+    # Update all_tokens_list
+    all_tokens_list.write(supply, token_id)
+
+    # Update all_tokens_index
+    all_tokens_index.write(token_id, supply)
+
+    # Update all_tokens_len
+    let (local new_supply: Uint256, _) = uint256_add(supply, Uint256(1, 0))
+    all_tokens_len.write(new_supply)
+    return ()
+end
+
+func _remove_token_from_all_tokens_enumeration{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(token_id: Uint256):
+    alloc_locals
+    let (supply: Uint256) = all_tokens_len.read()
+    let (local index_from_id: Uint256) = all_tokens_index.read(token_id)
+    let (local last_token_id: Uint256) = all_tokens_list.read(supply)
+
+    # Update all_tokens_list i.e. index n => token_id
+    all_tokens_list.write(index_from_id, last_token_id)
+
+    # Update all_tokens_index i.e. token_id => index n
+    all_tokens_index.write(last_token_id, index_from_id)
+
+    # Update totalSupply
+    let (local new_supply: Uint256) = uint256_sub(supply, Uint256(1, 0))
+    all_tokens_len.write(new_supply)
+    return ()
+end
+
+func _add_token_to_owner_enumeration{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(to: felt, token_id: Uint256):
+    alloc_locals
+    let (local length: Uint256) = ERC721_balanceOf(to) 
+    owned_tokens.write(to, length, token_id)
+    owned_tokens_index.write(token_id, length)
+    return ()
+end
+
+func _remove_token_from_owner_enumeration{
+        pedersen_ptr: HashBuiltin*, 
+        syscall_ptr: felt*, 
+        range_check_ptr
+    }(_from: felt, token_id: Uint256):
+    alloc_locals
+    let (local last_token_index: Uint256) = ERC721_balanceOf(_from)
+    # the index starts at zero therefore the user's last token index is their balance minus one
```

Comment:
why isn't this necessary in `_remove_token_from_all_tokens_enumeration`?

Reply:
I should have included a comment here regarding that. In the Solidity implementation, it's omitted here because of how rarely the last token minted is burnt and to save gas by not including an 'if' statement